db/memtable.cc

```diff
@@ -517,6 +520,58 @@ Status MemTable::Add(SequenceNumber s, ValueType type,
   size_t ts_sz = GetInternalKeyComparator().user_comparator()->timestamp_size();
   Slice key_without_ts = StripTimestampFromUserKey(key, ts_sz);
 
+  if (kv_prot_info != nullptr) {
+    Status verify_status;
+    KvProtectionInfo memtable_kv_prot_info;
+    Slice input(buf, encoded_len);
+    uint32_t verify_ikey_len = 0;
+    if (!GetVarint32(&input, &verify_ikey_len)) {
+      verify_status = Status::Corruption("Unable to parse internal key length");
+    }
+    if (verify_status.ok() && verify_ikey_len < 8 + ts_sz) {
+      verify_status = Status::Corruption("Internal key length too short");
+    }
+    if (verify_status.ok() && verify_ikey_len > input.size()) {
+      verify_status = Status::Corruption("Internal key length too long");
+    }
+    uint32_t verify_value_len = 0;
+    if (verify_status.ok()) {
+      const size_t verify_key_without_ts_len = verify_ikey_len - ts_sz - 8;
+      memtable_kv_prot_info.SetKeyChecksum(
+          GetSliceNPHash64(Slice(input.data(), verify_key_without_ts_len)));
+      input.remove_prefix(verify_key_without_ts_len);
+      if (ts_sz > 0) {
+        memtable_kv_prot_info.SetTimestampChecksum(
+            GetSliceNPHash64(Slice(input.data(), ts_sz)));
+      }
+      input.remove_prefix(ts_sz);
+      uint64_t verify_packed = DecodeFixed64(input.data());
+      input.remove_prefix(8);
+      uint64_t verify_seq;
+      ValueType verify_value_type;
+      UnPackSequenceAndType(verify_packed, &verify_seq, &verify_value_type);
+      memtable_kv_prot_info.SetSequenceNumber(verify_seq);
+      memtable_kv_prot_info.SetValueType(verify_value_type);
+      if (!GetVarint32(&input, &verify_value_len)) {
+        verify_status = Status::Corruption("Unable to parse value length");
+      }
+    }
+    if (verify_status.ok() && verify_value_len < input.size()) {
+      verify_status = Status::Corruption("Value length too short");
+    }
+    if (verify_status.ok() && verify_value_len > input.size()) {
+      verify_status = Status::Corruption("Value length too long");
+    }
+    if (verify_status.ok()) {
+      Slice verify_value = Slice(input.data(), verify_value_len);
+      memtable_kv_prot_info.SetValueChecksum(GetSliceNPHash64(verify_value));
+      verify_status = kv_prot_info->VerifyAgainst(memtable_kv_prot_info);
```

Comment:
Does it matter if you validate kv_prot_info against memtable_kv_prot_info or the other way around?  Which one do you care about the flags for?

Reply:
`kv_prot_info` may have more info because it accumulated some extra things like CF ID along its journey here. So it should be the `this` in `VerifyAgainst()`, as that verifies `this`'s protection info is a superset of `memtable_kv_prot_info`. (See above comment for a possible design change which may make things simpler.)